File: 50. Pow(x, n)/Python/Pow(x, n).py
class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if not n:
            return 1
        if n < 0:
            n = -n
            x = 1 / x
        if n % 2:
            return x * self.myPow(x * x, n // 2)
        return self.myPow(x * x, n // 2)

File: 50. Pow(x, n)/Java/Pow(x, n).java
public class Solution {
    public double myPow(double x, int n) {
        if (n == 0) {
            return 1;
        }
        if (n == Integer.MIN_VALUE) {
            n++;
            n = -n;
            x = 1 / x;
            return x * x * myPow(x * x, n / 2);
        }
        if (n < 0) {
            n = -n;
            x = 1 / x;
        }
        if (n % 2 == 0) {
            return myPow(x * x, n / 2);
        }
        return x * myPow(x * x, n / 2);
    }
}